### ISSUE TITLE
`nix build --dry-run` doesn't instantiate derivation

### ISSUE THREAD

**Issue by regnat** [NixOS/nix#3681]
**Describe the bug**

On master, `nix build --dry-run <any derivation that's not in the store>` fails with `error: store path name contains forbidden character`

**Steps To Reproduce**

With a nix built on top of current master (b2c8061b44b01a39d953d75db30e39bd8f300021):

```shell
# The spring package is marked as broken, so it's quite unlikely to be in the store. Otherwise choose something else
nix build --experimental-features nix-command --dry-run nixpkgs.spring
```

**Expected behavior**

This should print whatever needs to be done to realize the `nixpkgs.spring` derivation

**Additional context**

For nix 2.3.5, this doesn't work as expected either, but outputs:

```
don't know how to build these paths (may be caused by read-only store access):
  /nix/store/0k67dccvbjv9kdxx26v5vblh370xv9yw-spring-104.0.drv!out
```

I suspect (though that's a wild guess) that this failure is still there in master and causes the `forbidden character` error because the `/nix/store/…-spring-104.0.drv!out` is passed to the `StorePath` constructor (and fails because of the `!out`).

**Comment by edolstra** [comment]
Currently --dry-run gives this:

```
# nix build --dry-run
don't know how to build these paths (may be caused by read-only store access):
  /nix/store/whjbwqqc8x1ifslc15q19m2gn6dmbxmg-nix-2.4pre20210125_dirty.drv
```

which is not optimal either.

**Comment by Ma27** [comment]
@edolstra a fix seems to be in #3782.

**Comment by regnat** [comment]
I’m wondering whether the right solution for this is to make the store writeable in `--dry-run` or to keep all the derivations in memory (which would be nice, because it would allow for example to run `nix build --dry-run nixpkgs#hello --store https://cache.nixos.org`)

**Comment by edolstra** [comment]
Fixed in #3782.
